### ISSUE TITLE
dotnet-trace: Make it easier to get a dump of what EventSources and Events are actually used by a process

### ISSUE THREAD

**Issue by samsp-msft** [dotnet/runtime#77572]
Trying to configure dotnet-trace with which providers, levels and flags is complicated, and typically involves reading through a bunch of documentation to figure out what the possible providers are, and what events they are firing.

One way to solve this would be to have a mode for dotnet-trace which rather than collecting all the events, would do an enumeration for the event sources and the events that get fired - enumerating each event only once. You can then run your app in this mode and get a listing of the events that were fired. You can then feed this back into the command line for dotnet-trace as the mapping of the providers and levels that are needed for further analysis.

It could be something like:
```
dotnet-trace info [-h|--help]
    [-n, --name <name>] [--diagnostic-port] [-o|--output <trace-file-path>] [-p|--process-id <pid>]
    [--show-child-io]
    [-- <command>] (for target applications running .NET 5 or later) 
```

And give an output of something like:

```
=== Events Seen ===
Provider Name : Event Name : Event ID : Event Level (int) : Count
-----------------------------------------------------------------
System.Buffers.ArrayPoolEventSource : BufferRented : 1 : Verbose (5) : 79
System.Buffers.ArrayPoolEventSource : BufferAllocated : 2 : Informational (4) : 11
System.Buffers.ArrayPoolEventSource : BufferReturned : 3 : Verbose (5) : 80
System.Buffers.ArrayPoolEventSource : BufferTrimPoll : 5 : Informational (4) : 2
System.Buffers.ArrayPoolEventSource : BufferTrimmed : 4 : Informational (4) : 7
Microsoft-Extensions-Logging : FormattedMessage : 1 : LogAlways (0) : 33
Microsoft-Extensions-Logging : Message : 2 : LogAlways (0) : 33
Microsoft-Extensions-Logging : MessageJson : 5 : LogAlways (0) : 33
Microsoft-Diagnostics-DiagnosticSource : Message : 1 : Informational (4) : 1
Microsoft-Diagnostics-DiagnosticSource : NewDiagnosticListener : 10 : Informational (4) : 2
Microsoft-Diagnostics-DiagnosticSource : Event : 2 : Informational (4) : 2
Microsoft-Extensions-DependencyInjection : ServiceProviderBuilt : 7 : Informational (4) : 1
Microsoft-Extensions-DependencyInjection : ServiceProviderDescriptors : 8 : Informational (4) : 4
Microsoft-Extensions-DependencyInjection : CallSiteBuilt : 1 : Verbose (5) : 80
Microsoft-Extensions-DependencyInjection : ServiceResolved : 2 : Verbose (5) : 76
Microsoft-Extensions-DependencyInjection : DynamicMethodBuilt : 4 : Verbose (5) : 3
Microsoft-Extensions-DependencyInjection : ScopeDisposed : 5 : Verbose (5) : 1
System.Diagnostics.Eventing.FrameworkEventSource : ThreadPoolEnqueueWork : 30 : Verbose (5) : 12
System.Diagnostics.Eventing.FrameworkEventSource : ThreadPoolDequeueWork : 31 : Verbose (5) : 12
System.Diagnostics.Eventing.FrameworkEventSource : ThreadTransferSend : 150 : Informational (4) : 1
Microsoft.AspNetCore.Hosting : HostStart : 1 : Informational (4) : 1
Microsoft.AspNetCore.Hosting : HostStop : 2 : Informational (4) : 1
Private.InternalDiagnostics.System.Net.Sockets : Info : 4 : Informational (4) : 76
Private.InternalDiagnostics.System.Net.Sockets : ErrorMessage : 5 : Error (2) : 4
Dotnet-dev-certs : ListCertificatesStart : 1 : Verbose (5) : 1
Dotnet-dev-certs : DescribeFoundCertificates : 2 : Verbose (5) : 1
Dotnet-dev-certs : CheckCertificatesValidity : 3 : Verbose (5) : 1
Dotnet-dev-certs : DescribeValidCertificates : 4 : Verbose (5) : 1
Dotnet-dev-certs : DescribeInvalidCertificates : 5 : Verbose (5) : 1
Dotnet-dev-certs : ListCertificatesEnd : 6 : Verbose (5) : 1
System.Net.Sockets : AcceptStart : 4 : Informational (4) : 4
System.Net.Sockets : AcceptFailed : 6 : Error (2) : 4
System.Net.Sockets : AcceptStop : 5 : Informational (4) : 4
Microsoft-AspNetCore-Server-Kestrel : Configuration : 11 : LogAlways (0) : 1
System.Net.NameResolution : ResolutionStart : 1 : Informational (4) : 1
System.Net.NameResolution : ResolutionStop : 2 : Informational (4) : 1
Private.InternalDiagnostics.System.Net.NameResolution : Info : 4 : Informational (4) : 1
```

**Comment by samsp-msft** [comment]
This kind of information can be collected at runtime by modifying the code, for example https://github.com/samsp-msft/scrap/blob/main/EventSourceEnumerator.cs can be used to collect the data. In this case run the app within the body of a lambda, and it will dump out the events after the app has finished. 

For example:

```
EventSourceEnumerator.Run(async () =>
{
    var builder = WebApplication.CreateBuilder();
    var app = builder.Build();

    app.MapGet("/", () => "Error incorrect path: Use /TileService to access the service");
    app.Run();
});
```

But the whole point of dotnet-trace is to be able to use it outside of the process that is being instrumented.

**Comment by tommcdon** [comment]
@ldillonel @noahfalk @davmason

**Comment by noahfalk** [comment]
what if we did dynamic discovery of EventSources and static analysis to determine their events? Counting the number of events could be very high performance overhead for verbose events and other events may never occur depending on the workload? This would keep overhead minimal and make the list more complete.

On a side note, the demo is missing the runtime events due to a long standing sharp edge in how EventListener API is designed. No need to fix it though, it totally fulfills its demo role :) Sadly the EventListener starts enumerating event sources from within the base class constructor and that code runs prior to when the derived class constructor had a chance to register the event callback. Any EventSources that existed prior to the Listener being created didn't get observed.

**Comment by samsp-msft** [comment]
The "demo" was a quick mockup to validate a separate issue, and make sure specific events are being fired, but the need for it illustrated that trying to fish this information from documentation is not very efficient. 
There are 2 ways to view the problem:
- which events could be fired
- which events are actually being fired

The former could be done all by static analysis - pass in the app, and then use reflection to see what classes are present. But I feel that knowing what event sources are actually used has a lot of value. I show the actual events fired mostly so that the correct level can be selected for the filter - the demo should probably have ordered the events listed based on the level.
If there was a uniform way of knowing how the flags would be used, it would be helpful to show the events from them all, but I suspect that is beyond the scope of this.

Having a count of events fired was something that naturally fell out of the implementation of the prototype and is not necessary for the scenario, but if live enumeration is used, it helps to get an idea of what the volume is going to be expected to be.
